### ISSUE TITLE
 Sorry, couldn't find any help with that topic?

### ISSUE THREAD

**Issue by tassosblackg** [gleitz/howdoi#372]
howdoi version 2.0.14 gives this error no matter what search string I type

for example :
` howdoi format date bash`

Throughs = "Sorry, couldn't find any help with that topic?"

**Comment by kingpitka** [comment]
Same version, same problem.

**Comment by kingpitka** [comment]
Running _sanity_check() with no arguments also gives me:

Cache cleared successfully
Traceback:
Line 645, in _sanity_check
asset howdoi(google_args).encode('utf-8', 'ignore') != error_result
AssertionError
Line 647, in _sanity_check
Raise GoogleValidationError from exc
howdoi.errors.GoogleValidationError

**Comment by gleitz** [comment]
When this happens, I usually try two things: clearing the cache and potentially switching to another search engine if you think you've been rate limited.

```
howdoi format date bash -C
```

```
howdoi format date bash -e bing
```

Give those a try and let me know if either works for you.

**Comment by sunnysonx** [comment]
Same issue here. WSL (Ubuntu 18.04 LTS).

![image](https://user-images.githubusercontent.com/14160283/113907567-343ed900-97de-11eb-80ce-d3334ddd5000.png)

**Comment by kingpitka** [comment]
Neither command changed the results, sorry to say. Any diagnostic steps we could try running, besides the sanity check?

**Comment by gleitz** [comment]
Thanks for the update. There's an experimental `--explain` endpoint that is being built in https://github.com/gleitz/howdoi/pull/344

If you don't mind checking out that branch and running 

```
python -m howdoi.howdoi format date bash -C --explain
```

That might give some more insight into what is being blocked.

**Comment by kingpitka** [comment]
Okay, after checking #344 out and running:
```
>py -m howdoi.howdoi format date bash -C --explain
INFO: Version: 2.0.13
Cache cleared successfully
INFO: Fetching answers for query: format date bash
Sorry, couldn't find any help with that topic
(use --explain to learn why)
```

Seems like if --explain is tacked on, the reminder text at the end could be omitted. But still not working, sadly.

**Comment by kingpitka** [comment]
Tried running howdoi on a different computer, different network, same result. Tried running pysimplegui-howdoi and WebDoI, same result. Definitely feeling a little haunted at this point.

**Comment by gleitz** [comment]
Thanks for testing that. Does DDG fail as well?

```
howdoi format date bash -e duckduckgo
```

@KimaruThagna is working on a new version of the explainability that I'd like to use to see if we can debug this very puzzling issue. It's here -> https://github.com/gleitz/howdoi/pull/359

**Comment by kingpitka** [comment]
> @kingpitka that is spooooky. Let's see if we can get this explainability PR in a good place and use this situation as its first test.

Here's hoping, I absolutely love the idea of this program and hope to get it working soon. To add another level of complexity, "howdoi format date bash" works fantastically. Both with DuckDuckGo and the default engine.

> Can you access StackOverflow? Do you use a proxy or any kind of VPN?

Bit of an interesting situation, I've stripped my browser of internet access on this computer for a few less distractions (why howdoi sounds so ideal). All other programs can access things just fine, no proxy or VPN running, and requests connects immediately to StackOverflow. The browser situation shouldn't be bothering it at all.

**Comment by gleitz** [comment]
Oh, that's good to hear. So `howdoi format date bash` is working again? Are there queries that still aren't working?

`howdoi` definitely needs internet access to work - both to contact google and then get the answers from stackoverflow.

**Comment by kingpitka** [comment]
Most other queries still aren't working.

Doesn't work:
format time bash
print hello world python
for loop javascript
exit vim
for loop python
rename master main github
install package pip
comment html


Works:
format date bash
push commit github

I'll add if I find any more.

**Comment by tassosblackg** [comment]
I have tried almost everything updating, upgrading, reinstall it in different machines using Ubuntu18.04 or Manjaro 20.2 but the same condition doesn't do anything. every time the same message "Sorry..." I've tried also everything that is mentioned above but not any progress!

**Comment by gleitz** [comment]
Sorry, that sounds puzzling. Two things I can think of off the top of my head: you're either not able to reach the search engine chosen or not able to reach Stackoverflow. @KimaruThagna is working on a PR that will provide the `--explain` endpoint and will tell you which of these is failing. When he is ready with that PR we will post it here so you can give it a try.

**Comment by gleitz** [comment]
Hi all - thanks for being patient as we try to figure out what is wrong.

I'd like to try our the new explainability feature (https://github.com/gleitz/howdoi/pull/381) to see if we can track down the issue.

If you don't mind, try the following:

```
git clone https://github.com/gleitz/howdoi.git
git checkout origin/explainability-v2
python -m howdoi format date bash --explain -C
```

Here's the output I receive:

```
» python -m howdoi format date bash --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: stackoverflow.com links found on google: 1
INFO: Answers requested: 1, Starting at position: 1
INFO: Total answers returned: 1
# put current date as yyyy-mm-dd in $date
# -1 -> explicit current date, bash >=4.3 defaults to current time if not provided
# -2 -> start time for shell
printf -v date '%(%Y-%m-%d)T\n' -1

# put current date as yyyy-mm-dd HH:MM:SS in $date
printf -v date '%(%Y-%m-%d %H:%M:%S)T\n' -1

# to print directly remove -v flag, as such:
printf '%(%Y-%m-%d)T\n' -1
# -> current date printed to terminal
```

Hopefully the output will help us track down this spooky issue! 👻

**Comment by tassosblackg** [comment]
I've tried to follow the above comment's step @gleitz  to check and debug the issue

but when I've tried to run `python -m howdoi format date bash --explain -C`

I got a syntax error 

> ` Traceback (most recent call last):
>   File "/usr/lib/python2.7/runpy.py", line 163, in _run_module_as_main
>     mod_name, _Error)
> 
>   File "/usr/lib/python2.7/runpy.py", line 119, in _get_module_details
>     code = loader.get_code(mod_name)
> 
>   File "/usr/lib/python2.7/pkgutil.py", line 281, in get_code
>     self.code = compile(source, self.filename, 'exec')
> 
>   File "/home/tassos/howdoi/howdoi/howdoi.py", line 129
>     raise self.exception() from value_error
>-------------------------------- ^
> 
> SyntaxError: invalid syntax`
> 

*pointer is at (m) letter of from check the previous and after that lines, I didn't find something wrong?! Or did I missed something? weird?*

**Comment by gleitz** [comment]
Sorry, I should have mentioned that we only support Python 3.

**Comment by tassosblackg** [comment]
So, the command should be changed too 
`python3 -m howdoi format date bash --explain -C`

This is the output after running it with `python3` 

```
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine found no StackOverflow links
Sorry, couldn't find any help with that topic% 
```

**Comment by kingpitka** [comment]
Same result, except my cache clear looks slightly different:
```

py.exe -m howdoi format date bash --explain -C
INFO: Version: 2.0.14
←[92mCache cleared successfully←[0m
INFO: Fetching answers for query: format date bash
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine found no StackOverflow links
Sorry, couldn't find any help with that topic

````

**Comment by kingpitka** [comment]
I threw an extra `logging.info(html)` in `def _get_links()` and it may have spat a bit more of the answer out at me.

```
...
<h1>Before you continue to Google Search</h1>
...

Google uses <a href="https://policies.google.com/technologies/cookies?hl=en&amp;utm_source=ucb" target="_blank">cookies</a> and data to:<ul><li>Deliver and maintain services, like tracking outages and protecting against spam, fraud, and abuse</li><li>Measure audience engagement and site statistics to understand how our services are used</li></ul>If you agree, we’ll also use cookies and data to:<ul><li>Improve the quality of our services and develop new ones</li><li>Deliver and measure the effectiveness of ads</li><li>Show personalized content, depending on your settings</li><li>Show personalized or generic ads, depending on your settings, on Google and across the web</li></ul>For non-personalized content and ads, what you see may be influenced by things like the content you’re currently viewing and your location (ad serving is based on general location). Personalized content and ads can
...
```

**Comment by gleitz** [comment]
@kingpitka Aha! So there's some kind of agreement that Google is requesting. I wonder how we can avoid or accept that. 

What happens for bing (and duckduckgo)?

**Comment by tassosblackg** [comment]
`python3 -m howdoi format date bash -e duckduckgo --explain -C`

```
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20bash&t=hj&ia=web
INFO: stackoverflow.com links found on duckduckgo: 1
INFO: Answers requested: 1, Starting at position: 1
INFO: Fetching page: https://stackoverflow.com/questions/1401482/yyyy-mm-dd-format-date-in-shell-script
INFO: Total answers returned: 1
# put current date as yyyy-mm-dd in $date
# -1 -> explicit current date, bash >=4.3 defaults to current time if not provided
# -2 -> start time for shell
printf -v date '%(%Y-%m-%d)T\n' -1 

# put current date as yyyy-mm-dd HH:MM:SS in $date
printf -v date '%(%Y-%m-%d %H:%M:%S)T\n' -1 

# to print directly remove -v flag, as such:
printf '%(%Y-%m-%d)T\n' -1
# -> current date printed to terminal
```

**Comment by tassosblackg** [comment]
For Bing,
`python3 -m howdoi format date bash -e bing --explain -C`
```
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching bing with URL: https://www.bing.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine found no StackOverflow links
Sorry, couldn't find any help with that topic%  
```

**Comment by kingpitka** [comment]
Bing's HTML is unreadable, but it worked once then broke, DuckDuckGo worked twice and then threw a BlockError.

**Comment by rdxmb** [comment]
```
$ python3 -m howdoi format date bash --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine found no StackOverflow links
Sorry, couldn't find any help with that topic
```

```
$ python3 -m howdoi format date bash -e duckduckgo --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20bash&t=hj&ia=web
INFO: stackoverflow.com links found on duckduckgo: 1
INFO: Answers requested: 1, Starting at position: 1
INFO: Fetching page: https://stackoverflow.com/questions/1401482/yyyy-mm-dd-format-date-in-shell-script
INFO: Total answers returned: 1
# put current date as yyyy-mm-dd in $date
# -1 -> explicit current date, bash >=4.3 defaults to current time if not provided
# -2 -> start time for shell
printf -v date '%(%Y-%m-%d)T\n' -1 

# put current date as yyyy-mm-dd HH:MM:SS in $date
printf -v date '%(%Y-%m-%d %H:%M:%S)T\n' -1 

# to print directly remove -v flag, as such:
printf '%(%Y-%m-%d)T\n' -1
# -> current date printed to terminal
```

```
$ python3 -m howdoi format date bash -e bing --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching bing with URL: https://www.bing.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine found no StackOverflow links
Sorry, couldn't find any help with that topic
```


```
$ export HOWDOI_SEARCH_ENGINE=duckduckgo ; howdoi git revert last commit
Sorry, couldn't find any help with that topic

$ python3 -m howdoi format date bash -e duckduckgo --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20bash&t=hj&ia=web
INFO: stackoverflow.com links found on duckduckgo: 1
INFO: Answers requested: 1, Starting at position: 1
INFO: Fetching page: https://stackoverflow.com/questions/1401482/yyyy-mm-dd-format-date-in-shell-script
INFO: Total answers returned: 1
# put current date as yyyy-mm-dd in $date
# -1 -> explicit current date, bash >=4.3 defaults to current time if not provided
# -2 -> start time for shell
printf -v date '%(%Y-%m-%d)T\n' -1 

# put current date as yyyy-mm-dd HH:MM:SS in $date
printf -v date '%(%Y-%m-%d %H:%M:%S)T\n' -1 

# to print directly remove -v flag, as such:
printf '%(%Y-%m-%d)T\n' -1
# -> current date printed to terminal
$ python3 -m howdoi format date bash -e duckduckgo --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20bash&t=hj&ia=web
ERROR: Unable to find an answer because the search engine temporarily blocked the request. Please wait a few minutes or select a different search engine.
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/howdoi/howdoi/__main__.py", line 3, in <module>
    command_line_runner()
  File "/tmp/howdoi/howdoi/howdoi.py", line 782, in command_line_runner
    utf8_result = howdoi(args).encode('utf-8', 'ignore')
  File "/tmp/howdoi/howdoi/howdoi.py", line 605, in howdoi
    res = _get_answers(args)
  File "/tmp/howdoi/howdoi/howdoi.py", line 413, in _get_answers
    question_links = _get_links_with_cache(args['query'])
  File "/tmp/howdoi/howdoi/howdoi.py", line 392, in _get_links_with_cache
    links = _get_links(query)
  File "/tmp/howdoi/howdoi/howdoi.py", line 280, in _get_links
    raise BlockError('Temporary block by search engine')
howdoi.howdoi.BlockError: Temporary block by search engine
```

**Comment by gleitz** [comment]
Thanks for all the tests. Printing the the HTML when search results come back is a good idea. 

From that we can gather more of the block and error messages, and then start to fix or mitigate the issues. 

I wonder if we should start falling back to other search engines if, say, Google is blocked then just seamlessly use Bing.

**Comment by kingpitka** [comment]
@gleitz that would definitely fix some search engine issues without users ever noticing, and doesn't seem too tricky to implement.

**Comment by rdxmb** [comment]
> I wonder if we should start falling back to other search engines if, say, Google is blocked then just seamlessly use Bing.

An idea would be to have another environment var like `HOWDOI_SEARCH_ENGINE_AUTO_FALLBACK` , which is true by default. So experienced users have the possibility to disable that.

**Comment by hieuhc** [comment]
I got the same problems. The `bing` search engine works sometimes but `google` seems to block every of my requests. Do you guys know the reason of the blocking?

**Comment by gleitz** [comment]
If you open the Google URL you receive with the `--explain` flag, are you able to accept whatever agreement they want you to make? I would think that would then allow future searches to work.

**Comment by hieuhc** [comment]
I opened the Google URL in the incognito mode, it gave me the terms, I accepted but the query in terminal still didn't work. Seems like we will need a way to accept the Google terms for the request sent by the terminal. Similar thing happens with bing search, the only difference is that the term panel only appears after a few requests. I attach here captures of Google and Bing terms panel.

![Screen Shot 2021-04-17 at 8 46 01 PM](https://user-images.githubusercontent.com/8268223/115123777-df117d00-9fbe-11eb-8778-ec40a5608ee5.png)
![Screen Shot 2021-04-17 at 8 46 42 PM](https://user-images.githubusercontent.com/8268223/115123781-e20c6d80-9fbe-11eb-9561-409060a15659.png)

**Comment by mieel** [comment]
running a query with `-C` only works 1 out of 10 times, doesn't matter with `engine`.

**Comment by gleitz** [comment]
Thanks everyone for the info - I think we're getting closer to a resolution.

@mieel would you mind running the explain command listed in https://github.com/gleitz/howdoi/issues/372#issuecomment-819625100 and let me know what you get?

In https://github.com/gleitz/howdoi/commit/45a19fcf6c4ce7b26938ffef202502beed5fda36 I made two changes:
- First, with `--explain` we now print the HTML when the search engine does not return any StackOverflow links. This should make it much easier to catch the kind of blocked page errors we were seeing above. I created a separate issue https://github.com/gleitz/howdoi/issues/382 where we can address how to properly accept the terms and conditions (and I welcome any thoughts or PRs in that thread to solve the issue).
- Also, I noticed that when you search for a query that returns no results (or when you get a blocked error as mentioned above and no results come back), we store that result in the cache and will return it from then on. With this new PR we will look in the cache and if it's empty or has no results we'll go ahead and make another live query. This way we won't cache blocked responses which should resolve some of the issues folks are seeing.

If everyone could try the `--explain` endpoint again, I'd like to move everyone who is having "block"-related issues into https://github.com/gleitz/howdoi/issues/382 and we can keep discussing any other issues in this thread.

**Comment by mieel** [comment]
tried it again with the with the explainability-v2, it worked the first time, but the second time I got this error:
```
❯ python3 -m howdoi format date bash -e duckduckgo --explain -C
INFO: Version: 2.0.14
Cache cleared successfully
INFO: Fetching answers for query: format date bash
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20bash&t=hj&ia=web
ERROR: Unable to find an answer because the search engine temporarily blocked the request. Please wait a few minutes or select a different search engine.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/michielthai/howdoi/howdoi/__main__.py", line 3, in <module>
    command_line_runner()
  File "/Users/michielthai/howdoi/howdoi/howdoi.py", line 784, in command_line_runner
    utf8_result = howdoi(args).encode('utf-8', 'ignore')
  File "/Users/michielthai/howdoi/howdoi/howdoi.py", line 607, in howdoi
    res = _get_answers(args)
  File "/Users/michielthai/howdoi/howdoi/howdoi.py", line 416, in _get_answers
    question_links = _get_links_with_cache(args['query'])
  File "/Users/michielthai/howdoi/howdoi/howdoi.py", line 395, in _get_links_with_cache
    links = _get_links(query)
  File "/Users/michielthai/howdoi/howdoi/howdoi.py", line 280, in _get_links
    raise BlockError('Temporary block by search engine')
howdoi.howdoi.BlockError: Temporary block by search engine
```

**Comment by gleitz** [comment]
@mieel that sounds as expected. Two calls for the same query in quick succession with `-C` will likely result in a block error. 

If you remove the `-C` you should be all set.

**Comment by mieel** [comment]
Yeah I did that again it said Im still blocked, im not even spamming them.

I tried google:
got this from the first request
```
python3 -m howdoi format date bash -e google --explain
INFO: Version: 2.0.14
INFO: Fetching answers for query: format date bash
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&hl=en
INFO: Search engine google found no StackOverflow links, returned HTML is:
INFO: <!DOCTYPE html><html lang="en" dir="ltr"><head><style nonce="fmehhL+WZUsH50zzkJDVtQ">
a, a:link, a:visited, a:active, a:hover {
  color: #1a73e8;
  text-decoration: none;
}
body {
  font-family: Roboto,RobotoDraft,Helvetica,Arial,sans-serif;
  text-align: center;
  -ms-text-size-adjust: 100%;
  -moz-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
}
.box {
  border: 1px solid #dadce0;
  box-sizing: border-box;
  border-radius: 8px;
  margin: 24px auto 5px auto;
  max-width: 520px;
  padding: 24px;
}
h1 {
  color: #2c2c2c;
  font-size: 24px;
  hyphens: auto;
  margin: 24px 0;
}
p, .sub, .contentText {
  color: #5f6368;;
  font-size: 14px;
  line-height: 20px;
  letter-spacing: 0.2px;
  text-align: left;
}
.signin {
  text-align: right;
}
.image {
  display: block;
  margin: 14px auto;
}
a.button {
  color: #fff;
}
button, .button {
  border-radius: 4px;
  background-color: #1a73e8;
  border: none;
  color: #fff;
  cursor: pointer;
  font-family: Roboto,RobotoDraft,Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 500;
  height: 36px;
  margin: 0 4px;
  padding: 10px 24px;
}
input.button {
  -webkit-appearance: none;
}
.error {
  border: 2px solid #d93025;
  border-radius: 5px;
  color: #d93025;
  margin: auto;
  padding: 5px;
}
.footer {
  margin-top: 16px;
}
.footer a {
  margin-left: 24px;
  white-space: nowrap;
}
.footer a, .footer select, .footer input {
  color: #757575;
  font-size: 12px;
}
.languagepicker {
  display: inline-flex;
  margin: 2px 0;
}
#languageform,
#languageselect {
  margin: 0 2px;
}
#languageselect {
  color: #757575;
}
.detailspage {
  margin: 24px auto 0 auto;
  max-width: 700px;
}
.detailspage h1, .detailspage h2 {
  font-size: 22px
}
.setting {
  border: 1px solid #dadce0;
  box-sizing: border-box;
  border-radius: 8px;
  margin-bottom: 11px;
  padding: 24px 24px 20px 24px;
  text-align: right;
}
.sub {
  padding: 24px 24px 20px 24px;
}
hr {
  margin: 12px -24px 12px;
  border: 0;
  border-top: 1px solid #dadce0;
}
fieldset {
  border: none;
  padding: 0;
}
label {
  margin: 24px;
}
td {
  vertical-align: top;
}
.setting h2, .setting h3, h4 {
  color: #3c4043;
  margin: 0;
  text-align: left;
}
.yt-text {
  color: #5f6368;
  font-size: 14px;
  font-weight: 300;
}
.productLogoContainer {
  margin: auto;
  max-width: 360px;
}

/** Narrow screen (for example a mobile device). */
@media only screen and (max-width: 480px) {
  body {
    margin: 18px 14px;
  }
  .imgContainer {
    min-width: 72px;
  }
  .hideOnSmallWidth {
    display: none;
  }
  .footer form {
    margin-bottom: 3px;
  }
}

/** Normal/non-narrow screen. */
@media not screen and (max-width: 480px) {
  body {
    margin: 18px 25px;
  }
  .imgContainer {
    width: 51%;
  }
  .hideOnNormalWidth {
    display: none;
  }
  .footer form {
    display: inline;
  }
}
</style><title>Before you continue to Google Search</title><meta name="viewport" content="initial-scale=1, maximum-scale=5, width=device-width"><link rel="shortcut icon" href="//www.google.com/favicon.ico"></head><body><div class="signin"><a href="https://accounts.google.com/ServiceLogin?hl=en&amp;continue=https://www.google.com/search?q%3Dsite:stackoverflow.com%2520format%2520date%2520bash%26hl%3Den&amp;gae=cb-" class="button">Sign in</a></div><div class="box"><img src="//www.gstatic.com/images/branding/googlelogo/1x/googlelogo_color_68x28dp.png" srcset="//www.gstatic.com/images/branding/googlelogo/2x/googlelogo_color_68x28dp.png 2x" width="68" height="28" alt="Google"><div class="productLogoContainer"><img src="https://www.gstatic.com/ac/cb/scene_cookie_wall_search_v2.svg" alt="" width="100%" height="100%" class="image" aria-hidden="true"></div><h1>Before you continue to Google Search</h1><div class="contentText">Google uses <a href="https://policies.google.com/technologies/cookies?hl=en&utm_source=ucb" target="_blank">cookies</a> and data to:<ul><li>Deliver and maintain services, like tracking outages and protecting against spam, fraud, and abuse</li><li>Measure audience engagement and site statistics to understand how our services are used</li></ul>If you agree, we’ll also use cookies and data to:<ul><li>Improve the quality of our services and develop new ones</li><li>Deliver and measure the effectiveness of ads</li><li>Show personalized content, depending on your settings</li><li>Show personalized or generic ads, depending on your settings, on Google and across the web</li></ul>For non-personalized content and ads, what you see may be influenced by things like the content you’re currently viewing and your location (ad serving is based on general location). Personalized content and ads can be based on those things and your activity like Google searches and videos you watch on YouTube. Personalized content and ads include things like more relevant results and recommendations, a customized YouTube homepage, and ads that are tailored to your interests.<p>Click “Customize” to review options, including controls to reject the use of cookies for personalization and information about browser-level controls to reject some or all cookies for other uses. You can also visit g.co/privacytools anytime.</p></div><div><a href="https://consent.google.com/dl?continue=https://www.google.com/search?q%3Dsite:stackoverflow.com%2520format%2520date%2520bash%26hl%3Den&amp;gl=BE&amp;hl=en&amp;pc=srp&amp;src=1" class="button">Customize</a><form action="https://consent.google.com/s" method="POST" style="display:inline;"><input type="hidden" name="gl" value="BE"><input type="hidden" name="m" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&amp;hl=en"><input type="hidden" name="ca" value="r"><input type="hidden" name="x" value="8"><input type="hidden" name="v" value="cb.20210413-13-p0.en+FX+710"><input type="hidden" name="t" value="ADw3F8iYj5c_Zw9jDRXCCL9vaVjxwvYlPQ:1618771258717"><input type="hidden" name="hl" value="en"><input type="hidden" name="src" value="1"><input type="submit" value="I agree" class="button" aria-label="Agree to the use of cookies and other data for the purposes described"/></form></div></div><div class="footer"><form action="https://consent.google.com/ml" method="get"><select id="languageselect" name="hl"><option value="af">Afrikaans</option><option value="az">azərbaycan</option><option value="bs">bosanski</option><option value="ca">català</option><option value="cs">Čeština</option><option value="cy">Cymraeg</option><option value="da">Dansk</option><option value="de">Deutsch</option><option value="et">eesti</option><option value="en-GB">English&nbsp;(United Kingdom)</option><option value="en" selected>English&nbsp;(United States)</option><option value="es">Español&nbsp;(España)</option><option value="es-419">Español&nbsp;(Latinoamérica)</option><option value="eu">euskara</option><option value="fil">Filipino</option><option value="fr-CA">Français&nbsp;(Canada)</option><option value="fr">Français&nbsp;(France)</option><option value="ga">Gaeilge</option><option value="gl">galego</option><option value="hr">Hrvatski</option><option value="id">Indonesia</option><option value="zu">isiZulu</option><option value="is">íslenska</option><option value="it">Italiano</option><option value="sw">Kiswahili</option><option value="lv">latviešu</option><option value="lt">lietuvių</option><option value="hu">magyar</option><option value="ms">Melayu</option><option value="nl">Nederlands</option><option value="no">norsk</option><option value="uz">o‘zbek</option><option value="pl">polski</option><option value="pt-BR">Português&nbsp;(Brasil)</option><option value="pt-PT">Português&nbsp;(Portugal)</option><option value="ro">română</option><option value="sq">shqip</option><option value="sk">Slovenčina</option><option value="sl">slovenščina</option><option value="sr-Latn">srpski (latinica)</option><option value="fi">Suomi</option><option value="sv">Svenska</option><option value="vi">Tiếng Việt</option><option value="tr">Türkçe</option><option value="el">Ελληνικά</option><option value="be">беларуская</option><option value="bg">български</option><option value="ky">кыргызча</option><option value="kk">қазақ тілі</option><option value="mk">македонски</option><option value="mn">монгол</option><option value="ru">Русский</option><option value="sr">српски</option><option value="uk">Українська</option><option value="ka">ქართული</option><option value="hy">հայերեն</option><option value="iw">עברית</option><option value="ur">اردو</option><option value="ar">العربية</option><option value="fa">فارسی</option><option value="am">አማርኛ</option><option value="ne">नेपाली</option><option value="mr">मराठी</option><option value="hi">हिन्दी</option><option value="as">অসমীয়া</option><option value="bn">বাংলা</option><option value="pa">ਪੰਜਾਬੀ</option><option value="gu">ગુજરાતી</option><option value="or">ଓଡ଼ିଆ</option><option value="ta">தமிழ்</option><option value="te">తెలుగు</option><option value="kn">ಕನ್ನಡ</option><option value="ml">മലയാളം</option><option value="si">සිංහල</option><option value="th">ไทย</option><option value="lo">ລາວ</option><option value="my">မြန်မာ</option><option value="km">ខ្មែរ</option><option value="ko">한국어</option><option value="ja">日本語</option><option value="zh-CN">简体中文</option><option value="zh-TW">繁體中文</option><option value="zh-HK">繁體中文&nbsp;(香港)</option></select><input type="hidden" name="oldhl" value="en"><input type="hidden" name="gl" value="BE"><input type="hidden" name="m" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20format%20date%20bash&amp;hl=en"><input type="hidden" name="src" value="1"><input type="submit" value="Change language"></form><a href="https://policies.google.com/privacy?hl=en">Privacy Policy</a> <a href="https://policies.google.com/terms?hl=en">Terms of Service</a></div></body></html>
Sorry, couldn't find any help with that topic%

 ~/howdoi  @45a19fcf ▓▒░                                                                                                               ░▒▓ 20:40:58
❯ python3 -m howdoi format date bash -e google --explain
INFO: Version: 2.0.14
INFO: Using cached response (add -C to clear the cache)
Sorry, couldn't find any help with that topic%

```

**Comment by gleitz** [comment]
For the folks who are getting blocked, are you in Europe? I read this might be GDPR related.

https://stackoverflow.com/questions/64846902/how-to-get-rid-of-the-google-cookie-pop-up-with-my-selenium-automation
https://stackoverflow.com/questions/64927752/google-consent-automatic-accepting

I found these articles that refer to bypassing consent.google.com, but I don't know how to reliably trigger the page.

**Comment by gleitz** [comment]
Explainability and a bunch of other goodies have been released in version `2.0.15` so feel free to upgrade using `pip`

**Comment by mieel** [comment]
Yep EU here (Belgium)

**Comment by kingpitka** [comment]
And the spooky mystery is laid to rest. I'll try connecting through a VPN later today and see what happens.

**Comment by rdxmb** [comment]
Yes, Europe (Germany)

```
$ howdoi git revert last commit
Sorry, couldn't find any help with that topic (use --explain to learn why)
$ howdoi --version
2.0.15
```

**Comment by hieuhc** [comment]
> For the folks who are getting blocked, are you in Europe? I read this might be GDPR related.
> 
> https://stackoverflow.com/questions/64846902/how-to-get-rid-of-the-google-cookie-pop-up-with-my-selenium-automation
> https://stackoverflow.com/questions/64927752/google-consent-automatic-accepting
> 
> I found these articles that refer to bypassing consent.google.com, but I don't know how to reliably trigger the page.

Yep I think it is related. I base in Europe too.

**Comment by flaviumarinescu** [comment]
If I try from my own computer (Europe) - not working - same issues as you mentioned.
I tried from google colab, it works with multiple requests. The IP address seems to be from Taiwan.


![Screenshot_20210422_00:11:54](https://user-images.githubusercontent.com/22303534/115621646-7ba98700-a2ff-11eb-831e-3cdc4e1fcce5.jpg)

**Comment by gleitz** [comment]
For those in Europe, is there a particular search engine that seems to work? Google and Bing are probably tracking you, but DDG might now?

Can anyone confirm if `-e duckduckgo` works better?

Then we can make

`alias h='howdoi --engine duckduckgo'`

and use with 

`h format date bash` to at least keep people working smoothly until we get a PR that will be able to send the right acceptance of the confirmation pages.

**Comment by rdxmb** [comment]
On my remote machine (Europe)

```
$ ./howdoi --engine duckduckgo git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
$ ./howdoi --engine bing git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
$ ./howdoi --engine google git undo last commit
Sorry, couldn't find any help with that topic (use --explain to learn why)
```
I can run all these multiple times (about 20x), no changes.

----------------------------------------------------------------------------

On my local machine (on which I was blocked last time)
```
# howdoi --engine bing git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine bing git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine bing git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine bing git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# howdoi --engine google git undo last commit
$ git commit -m "Something terribly misguided" # (0: Your Accident)
$ git reset HEAD~                              # (1)
[ edit files as necessary ]                    # (2)
$ git add .                                    # (3)
$ git commit -c ORIG_HEAD                      # (4)
# 

```

I am wondering what's going on here. I can also run this with google and `watch`, no problems.

**Comment by hieuhc** [comment]
Duckduckgo works for the first request, then consecutive requests are blocked by the search engine with `403 Client Error: Forbidden`

```
INFO: Fetching answers for query: format date in python
INFO: Searching duckduckgo with URL: https://duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20in%20python&t=hj&ia=web
INFO: http erro: HTTPError('403 Client Error: Forbidden for url: https://html.duckduckgo.com/html?q=site:stackoverflow.com%20format%20date%20in%20python&t=hj&ia=web')
INFO: result is None
ERROR: Unable to find an answer because the search engine temporarily blocked the request. Please wait a few minutes or select a different search engine.
Traceback (most recent call last):
  File "/howdoi/howdoi/howdoi.py", line 607, in howdoi
    res = _get_answers(args)
  File "/howdoi/howdoi/howdoi.py", line 415, in _get_answers
    question_links = _get_links_with_cache(args['query'])
  File "/howdoi/howdoi/howdoi.py", line 394, in _get_links_with_cache
    links = _get_links(query)
  File "/howdoi/howdoi/howdoi.py", line 281, in _get_links
    raise BlockError('Temporary block by search engine')
howdoi.howdoi.BlockError: Temporary block by search engine
```

**Comment by kingpitka** [comment]
Hey, could the Europeans in this thread try out my fork over at #393? I attached a small consent cookie to the request, which seems to have bypassed the Google GDPR Page for me :)

**Comment by gleitz** [comment]
> Hey, could the Europeans in this thread try out my fork over at #393? I attached a small consent cookie to the request, which seems to have bypassed the Google GDPR Page for me :)

And that fixes this issue. Thanks everyone for all the help.

**Comment by rdxmb** [comment]
Seems we have the problem again:

```
howdoi bash case --explain
WARNING:root:no boto3 module found
INFO:root:Version: 2.0.20
INFO:root:Fetching answers for query: bash case
INFO:root:Using cached links
INFO:root:No StackOverflow links found in cached search engine results - will make live query
INFO:root:Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&hl=en
INFO:root:Search engine google found no StackOverflow links, returned HTML is:
INFO:root:<!DOCTYPE html><html lang="en" dir="ltr"><head><style nonce="Sht4HiiWlXUwm2U_LjgZYg">
a, a:link, a:visited, a:active, a:hover {
  color: #1a73e8;
  text-decoration: none;
}
body {
  font-family: Roboto,Helvetica,Arial,sans-serif;
  text-align: center;
  -ms-text-size-adjust: 100%;
  -moz-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
}
.box {
  border: 1px solid #dadce0;
  box-sizing: border-box;
  border-radius: 8px;
  margin: 24px auto 5px auto;
  max-width: 800px;
  padding: 24px;
}
.youtubeContainerUIModernization,
.boxUIModernization {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 800px;
}
.signInContainerUIModernization {
    display: flex;
    justify-content: flex-end;
}
h1 {
  color: #2c2c2c;
  font-size: 24px;
  hyphens: auto;
  margin: 24px 0;
}
.icaCallout {
  background-color: #f8f9fa;
  padding: 12px 16px;
  border-radius: 10px;
  margin-bottom: 10px;
}
.icaCalloutUIModernization {
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 18px;
  letter-spacing: 0.2px;
  background-color: #f8f9fa;
  padding: 12px 24px;
  border-radius: 12px;
  margin-bottom: 16px;
  text-align: left;
}
.icaCalloutUIModernization > a {
  color: #065fd4;  /* youtube dark-blue color */
}
.subUIModernization,
.contentTextUIModernization,
.contentTextUIModernization > p {
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 400;
  font-size: 14px;
  letter-spacing: 0;
  line-height: 18px;
  text-align: left;
}
.contentTextUIModernization > a,
.contentTextUIModernization > p > a {
  color: #065fd4;
}
.contentTextUIModernization > p {
  margin: 16px 0;
}
.contentTextUIModernization > ul {
  padding-left: 36px;
  margin: 16px 0;
}
p, .sub, .contentText, .icaCallout {
  color: #5f6368;;
  font-size: 14px;
  line-height: 20px;
  letter-spacing: 0.2px;
  text-align: left;
}
.signin {
  text-align: right;
}
.signInIconContainer {
  width: 24px;
  height: 24px;
  margin: 0 6px 0 -6px;
}
.signInIcon {
  fill: #065fd4; /* dark-blue-color */
}
.signinUIModernization {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  text-align: right;
}
.bulletUIModernization {
  line-height: 18px;
  margin: 8px 0;
}
.signinButtonDisplayUIModernization {
  display: flex;
  box-sizing: border-box;
  align-items: center;
  flex-direction: row;
  justify-content: flex-start;
}
.horizontalContainerEndUIModernization {
  display: flex;
  align-items: center;
  justify-content: flex-end;
}
.cardHeaderUIModernization {
  display: flex;
  flex-direction: row;
  align-items: center;
}
.logoCenter {
  justify-content: center;
}
.logoUIModernization {
  margin-right: auto;
}
.youTubeHeaderUIModernization {
 padding: 20px 24px 0;
}
.saveButtonContainer,
.saveButtonContainerNarrowScreen {
  width: 100%;
  margin-top: 12px;
}
.saveButtonContainerNarrowScreenUIModernization {
  width: 100%;
}
.customButtonContainer {
  height: 20px;
  padding: 8px 0;
  margin-top: 8px;
}
.navigationCtasYoutube {
  text-align: right;
}
a.button {
  color: #fff;
}
a.hairlinebutton,
a.nolinebutton{
  color: #1a73e8;
}
.navigationCtasYoutube a.hairlinebutton {
  border-color: #1a73e8;
}
button, .button {
  background-color: #1a73e8;
  border: none;
  color: #fff;
}
input.button {
  -webkit-appearance: none;
}
.basebutton {
  border-radius: 4px;
  cursor: pointer;
  font-family: Roboto,Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 500;
  height: 36px;
  margin: 12px 4px 0;
  padding: 8px 24px;
}
.hairlinebutton {
  background-color: #fff;
  border-width: 1px;
  border-color: #dadce0;
  border-style: solid;
  max-height: 15px;
}
.basebuttonUIModernization {
  outline: none;
  text-align: center;
  vertical-align: middle;
  line-height: 18px;
  cursor: pointer;
  font-family: Roboto, Arial, sans-serif;
  font-size: 14px;
  font-weight: 500;
  font-style: normal;
  height: 36px;
  min-width: 72px;
  border-radius: 18px;
  padding: 0 16px;
}
.signInButtonUIModernization {
  background-color: #fff;
  border: 1px solid #e5e5e5;
  color: #065fd4;
  }
.signInButtonUIModernization:focus-visible {
  background-color: #fff;
  border: 1px solid #065fd4;
  color: #065fd4;
}
.signInButtonUIModernization:hover {
  background-color: #f2f8ff;
  border: 1px solid transparent;
  color: #065fd4;
}
.signInButtonUIModernization:hover:focus-visible {
  background-color: #f2f8ff;
  border: 1px solid #065fd4;
  color: #065fd4;
}
.signInButtonUIModernization:active,
.signInButtonUIModernization:active:hover,
.signInButtonUIModernization:active:focus-visible {
  background-color: #e5e5e5;
  border: 1px solid transparent;
  color: #065fd4;
}
.signinUIModernization a,
.signinUIModernization a:link,
.signinUIModernization a:visited,
.signinUIModernization a:active,
.signinUIModernization a:hover,
.signinUIModernization a:focus-visible {
  color: #065fd4;
}
.hairlinebuttonUIModernization {
  background-color: #0f0f0f;
  border: 2px solid transparent;
  color: #fff;
  }
.hairlinebuttonUIModernization:focus-visible {
  background-color: #fff;
  border: 2px solid #0f0f0f;
  color: #0f0f0f;
}
.hairlinebuttonUIModernization:hover {
  background-color: #272727;
  border: 2px solid transparent;
  color: #fff;
}
.hairlinebuttonUIModernization:hover:focus-visible {
  background-color: #e7e7e7;
  border: 2px solid transparent;
  color: #0f0f0f;
}
.hairlinebuttonUIModernization:active,
.hairlinebuttonUIModernization:active:hover,
.hairlinebuttonUIModernization:active:focus-visible {
  background-color: #3f3f3f;
  border: 2px solid transparent;
  color: #fff;
}
.customizeButtonContainerUIModernization {
  display: flex;
  text-align: center;
}
.customizeButtonUIModernization {
  box-sizing: border-box;
  background-color: #f2f2f2;
  border: 2px solid transparent;
  text-align: center;
}
.customizeButtonUIModernization:focus-visible {
  background-color: #fff;
  border: 2px solid #0f0f0f;
}
.customizeButtonUIModernization:hover {
  background-color: #e6e6e6;
  border: 2px solid transparent;
}
.customizeButtonUIModernization:active,
.customizeButtonUIModernization:active:hover,
.customizeButtonUIModernization:active:focus-visible {
  background-color: #cfcfcf;
  border: 2px solid transparent;
}
a.basebuttonUIModernization.customizeButtonUIModernization,
a.basebuttonUIModernization.customizeButtonUIModernization:link,
a.basebuttonUIModernization.customizeButtonUIModernization:visited,
a.basebuttonUIModernization.customizeButtonUIModernization:active,
a.basebuttonUIModernization.customizeButtonUIModernization:focus {
  color: #0f0f0f;
  text-decoration: none;
  padding: 8px 16px;
}
.saveButtonUIModernization {
  text-align: center;
}
.detailsButtonUIModernization {
  margin-top: 16px;
  min-width: 220px;
}
.error {
  border: 2px solid #d93025;
  border-radius: 5px;
  color: #d93025;
  margin: auto;
  padding: 5px;
}
.box a:hover, .boxUIModernization a:hover, .detailspage a:hover {
  text-decoration: underline;
}
.footer {
  margin-top: 16px;
}
.footerLinks {
  display: flex;
  justify-content: center;
  align-items: center;
}
.footerUIModernization {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-top: 16px;
}
.footer a {
  margin-left: 24px;
  white-space: nowrap;
}
.footer a, .footer select, .footer input {
  color: #757575;
  font-size: 12px;
}
.footerUIModernization a {
  margin-left: 8px;
  padding: 0 8px;
}
.footerUIModernization a,
.footerUIModernization input,
.footerUIModernization select {
  color: #606060;  /* grey5 color - youtube secondary color */
  font-family: Roboto;
  font-size: 12px;
  font-style: normal;
  font-weight: 500;
}
.footerUIModernization form {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}
.languagepicker {
  display: inline-flex;
  margin: 2px 0;
}
#languageform,
#languageselect {
  margin: 0 2px;
}
#languageselect {
  color: #757575;
}
.detailspage {
  margin: 24px auto 0 auto;
  max-width: 700px;
}
.detailspage h1, .detailspage h2 {
  font-size: 22px
}
.setting {
  border: 1px solid #dadce0;
  box-sizing: border-box;
  border-radius: 8px;
  margin-bottom: 11px;
  padding: 24px 24px 20px 24px;
  text-align: right;
}
.settingUIModernization {
  border: 1px solid #dadce0;
  box-sizing: border-box;
  border-radius: 8px;
  margin-bottom: 16px;
  padding: 24px;
  text-align: right;
}
.sub {
  padding: 24px 24px 20px 24px;
}
hr {
  margin: 12px -24px 12px;
  border: 0;
  border-top: 1px solid #dadce0;
}
fieldset {
  border: none;
  padding: 0;
}
label {
  margin: 24px;
}
td {
  vertical-align: top;
}
.setting h2, .setting h3, h4 {
  color: #3c4043;
  margin: 0;
  text-align: left;
}
.yt-text {
  color: #5f6368;
  font-size: 14px;
  font-weight: 300;
}
.detailspageUIModernization {
  margin: 0 auto;
  max-width: 700px;
}
.subUIModernization > ul > li,
.detailspageUIModernization p{
  color: #606060;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 400;
  font-size: 14px;
  line-height: 18px;
}
.detailspageUIModernization > p,
.settingUIModernization p,
.subUIModernization > p,
.subUIModernization > ul {
  margin: 16px 0;
}
.subUIModernization p:last-of-type {
  margin-bottom: 0;
}
.settingUIModernization fieldset {
  margin: 0;  /* removing the default margin of the fieldset element */
}
.subUIModernization > ul {
  padding-left: 36px;
}
.subUIModernization > ul > li {
  line-height: 18px;
  margin: 8px 0;
}
.detailspageUIModernization h1{
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 700;
  font-size: 24px;
  line-height: 30px;
}
.settingUIModernization h2 {
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 700;
  font-size: 18px;
  line-height: 22px;
  margin: 0;
  text-align: left;
}
.detailspageUIModernization h4 {
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 500;
  font-size: 16px;
  line-height: 20px;
  margin: 0;
  text-align: left;
}
.subUIModernization > a,
.subUIModernization > p > a,
.detailspageUIModernization > a,
.detailspageUIModernization > p > a {
  color: #065fd4;
}
.detailspageUIModernization a:hover {
  text-decoration: underline;
}
.subUIModernization {
  padding: 0 24px;
}
.ytMainHeaderUIModernization {
  color: #0f0f0f;
  font-family: Roboto, Arial, sans-serif;
  font-weight: 700;
  font-size: 24px;
  line-height: 30px;
}
.ytTextUIModernization {
  color: #606060;
  font-family: 'Google Sans';
  font-weight: 400;
  font-size: 14px;
  line-height: 28px;
  margin-left: 8px;
  vertical-align: middle;
  display: inline-block;
}
.productLogoContainer {
  margin: auto;
  max-width: 360px;
}
.ytLogoUIModernization {
  display: flex;
  flex-direction: row;
  justify-content: start;
  align-items: center;
}
.ytLogoImageUIModernization {
  vertical-align: middle;
  display: inline-block;
  margin-bottom: 8px;
}
.horizontalSeparator {
  margin: 12px -24px;
}
.horizontalSeparatorUIModernization {
  margin: 16px -24px;
}
/** Narrow screen (for example a mobile device). */
@media only screen and (max-width: 480px) {
  body {
    margin: 18px 14px;
  }
  .bodyUIModernization {
    margin: 24px 16px;
  }
  .ytLogoUIModernization {
    flex-direction: column;
    align-items: flex-start;
  }
  .ytLogoImageUIModernization {
    margin-bottom: 0;
  }
  .cardHeaderUIModernization {
    justify-content: space-between;
  }
  .ytTextUIModernization {
    margin-left: 0;
    margin-top: -4px;
  }
  .settingUIModernization {
    padding: 16px;
  }
  .settingUIModernization p {
    margin-top: 8px;
  }
  .subUIModernization {
    padding: 0;
  }
  .horizontalSeparatorUIModernization {
    margin: 16px 0;
  }
  .footer form {
    margin-bottom: 3px;
  }
  .imgContainer {
    min-width: 96px;
  }
  button, .button {
    width:100%;
  }
  .basebutton {
    margin: 12px 0 0;
  }
  .hideOnSmallWidth {
    display: none;
  }
  .saveButtonContainerNarrowScreen,
  .saveButtonContainerNarrowScreenUIModernization {
    display: inline-block;
  }
  .saveButtonContainer {
    display: none;
  }
  .saveButtonContainerUIModernization {
    display: none;
  }
  .navigationCtasYoutube a.hairlinebutton {
    display: block;
    text-align: center;
  }
  .detailsButtonUIModernization,
  .saveButtonUIModernization {
    display: block;
    box-sizing: border-box;
    width:100%;
    margin-bottom: 16px;
  }
  .customizeButtonUIModernization {
    display: block;
    box-sizing: border-box;
    width:100%;
  }
}

@media only screen and (max-width: 380px) {
  .imgContainer {
    min-width: 72px;
  }
}

/** Normal/non-narrow screen. */
@media not screen and (max-width: 480px) {
  body {
    margin: 18px 25px;
  }
  .bodyUIModernization {
    margin: 24px;
  }
  .boxUIModernization {
    border: 1px solid #dadce0; /* grey-300 color */
    border-radius: 12px;
    padding: 24px;
  }
  .signInContainerUIModernization {
    margin-bottom: 24px;
  }
  .footer form {
    display: inline;
  }
  .imgContainer {
    width: 51%;
  }
  button, .button {
    width: auto;
  }
  .searchButton {
    min-width: 182px;
  }
  .detailsButton {
    min-width: 220px;
  }
  .hideOnNormalWidth {
    display: none;
  }
  .saveButtonContainer {
    display: inline-block;
  }
  .saveButtonContainerNarrowScreen,
  .saveButtonContainerNarrowScreenUIModernization {
    display: none;
  }
  .saveButtonContainerUIModernization {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
  }
  .saveButtonUIModernization {
    margin-left: 8px;
  }
}
</style><title>Before you continue to Google Search</title><meta name="viewport" content="initial-scale=1, maximum-scale=5, width=device-width"><link rel="shortcut icon" href="//www.google.com/favicon.ico"></head><body class=""><div><div class="signin"><a href="https://accounts.google.com/ServiceLogin?hl=en&amp;continue=https://www.google.com/search?q%3Dsite:stackoverflow.com%2520bash%2520case%26hl%3Den&amp;gae=cb-none" class="basebutton hairlinebutton hideOnSmallWidth">Sign in</a></div></div><div class=""><div class="box"><div class="cardHeaderUIModernization logoCenter"><img src="//www.gstatic.com/images/branding/googlelogo/1x/googlelogo_color_68x28dp.png" srcset="//www.gstatic.com/images/branding/googlelogo/2x/googlelogo_color_68x28dp.png 2x" width="68" height="28" alt="Google"><div class="signin"><a href="https://accounts.google.com/ServiceLogin?hl=en&amp;continue=https://www.google.com/search?q%3Dsite:stackoverflow.com%2520bash%2520case%26hl%3Den&amp;gae=cb-none" class="basebutton hairlinebutton hideOnNormalWidth">Sign in</a></div></div><h1>Before you continue to Google</h1><div class="contentText"><p>We use <a href="https://policies.google.com/technologies/cookies?hl=en&utm_source=ucb" target="_blank">cookies</a> and data to</p><ul><li>Deliver and maintain Google services</li><li>Track outages and protect against spam, fraud, and abuse</li><li>Measure audience engagement and site statistics to understand how our services are used and enhance the quality of those services</li></ul><p>If you choose to “Accept all,” we will also use cookies and data to</p><ul><li>Develop and improve new services</li><li>Deliver and measure the effectiveness of ads</li><li>Show personalized content, depending on your settings</li><li>Show personalized ads, depending on your settings</li></ul><p>If you choose to “Reject all,” we will not use cookies for these additional purposes.</p><p>Non-personalized content is influenced by things like the content you’re currently viewing, activity in your active Search session, and your location. Non-personalized ads are influenced by the content you’re currently viewing and your general location. Personalized content and ads can also include more relevant results, recommendations, and tailored ads based on past activity from this browser, like previous Google searches. We also use cookies and data to tailor the experience to be age-appropriate, if relevant.</p><p>Select “More options” to see additional information, including details about managing your privacy settings. You can also visit g.co/privacytools at any time.</p></div><div class="saveButtonContainer"><form action="https://consent.google.com/save" method="POST" style="display:inline;"><input type="hidden" name="bl" value="boq_identityfrontenduiserver_20231203.09_p0"><input type="hidden" name="x" value="8"><input type="hidden" name="gl" value="DE"><input type="hidden" name="m" value="0"><input type="hidden" name="app" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&amp;hl=en"><input type="hidden" name="hl" value="en"><input type="hidden" name="uxe" value="none"><input type="hidden" name="cm" value="4"><input type="hidden" name="set_eom" value="true"><input type="submit" value="Reject all" class="basebutton button searchButton" aria-label="Reject all"/></form><form action="https://consent.google.com/save" method="POST" style="display:inline;"><input type="hidden" name="bl" value="boq_identityfrontenduiserver_20231203.09_p0"><input type="hidden" name="x" value="8"><input type="hidden" name="gl" value="DE"><input type="hidden" name="m" value="0"><input type="hidden" name="app" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&amp;hl=en"><input type="hidden" name="hl" value="en"><input type="hidden" name="uxe" value="none"><input type="hidden" name="cm" value="4"><input type="hidden" name="set_eom" value="false"><input type="hidden" name="set_sc" value="true"><input type="hidden" name="set_aps" value="true"><input type="submit" value="Accept all" class="basebutton button searchButton" aria-label="Accept all"/></form></div><div class="saveButtonContainerNarrowScreen"><form action="https://consent.google.com/save" method="POST" style="display:block;"><input type="hidden" name="bl" value="boq_identityfrontenduiserver_20231203.09_p0"><input type="hidden" name="x" value="8"><input type="hidden" name="gl" value="DE"><input type="hidden" name="m" value="0"><input type="hidden" name="app" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&amp;hl=en"><input type="hidden" name="hl" value="en"><input type="hidden" name="uxe" value="none"><input type="hidden" name="cm" value="4"><input type="hidden" name="set_eom" value="false"><input type="hidden" name="set_sc" value="true"><input type="hidden" name="set_aps" value="true"><input type="submit" value="Accept all" class="basebutton button searchButton" aria-label="Accept all"/></form><form action="https://consent.google.com/save" method="POST" style="display:block;"><input type="hidden" name="bl" value="boq_identityfrontenduiserver_20231203.09_p0"><input type="hidden" name="x" value="8"><input type="hidden" name="gl" value="DE"><input type="hidden" name="m" value="0"><input type="hidden" name="app" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&amp;hl=en"><input type="hidden" name="hl" value="en"><input type="hidden" name="uxe" value="none"><input type="hidden" name="cm" value="4"><input type="hidden" name="set_eom" value="true"><input type="submit" value="Reject all" class="basebutton button searchButton" aria-label="Reject all"/></form></div><div class="customButtonContainer"><a href="https://consent.google.com/dl?continue=https://www.google.com/search?q%3Dsite:stackoverflow.com%2520bash%2520case%26hl%3Den&amp;gl=DE&amp;hl=en&amp;cm=4&amp;pc=srp&amp;uxe=none&amp;src=1" class="basebutton nolinebutton">More options</a></div></div><div class="footer"><form action="https://consent.google.com/ml" method="get"><select id="languageselect" name="hl"><option value="af">Afrikaans</option><option value="az">azərbaycan</option><option value="bs">bosanski</option><option value="ca">català</option><option value="cs">Čeština</option><option value="cy">Cymraeg</option><option value="da">Dansk</option><option value="de">Deutsch</option><option value="et">eesti</option><option value="en-GB">English&nbsp;(United Kingdom)</option><option value="en" selected>English&nbsp;(United States)</option><option value="es">Español&nbsp;(España)</option><option value="es-419">Español&nbsp;(Latinoamérica)</option><option value="eu">euskara</option><option value="fil">Filipino</option><option value="fr-CA">Français&nbsp;(Canada)</option><option value="fr">Français&nbsp;(France)</option><option value="ga">Gaeilge</option><option value="gl">galego</option><option value="hr">Hrvatski</option><option value="id">Indonesia</option><option value="zu">isiZulu</option><option value="is">íslenska</option><option value="it">Italiano</option><option value="sw">Kiswahili</option><option value="lv">latviešu</option><option value="lt">lietuvių</option><option value="hu">magyar</option><option value="ms">Melayu</option><option value="nl">Nederlands</option><option value="no">norsk</option><option value="uz">o‘zbek</option><option value="pl">polski</option><option value="pt-BR">Português&nbsp;(Brasil)</option><option value="pt-PT">Português&nbsp;(Portugal)</option><option value="ro">română</option><option value="sq">shqip</option><option value="sk">Slovenčina</option><option value="sl">slovenščina</option><option value="sr-Latn">srpski (latinica)</option><option value="fi">Suomi</option><option value="sv">Svenska</option><option value="vi">Tiếng Việt</option><option value="tr">Türkçe</option><option value="el">Ελληνικά</option><option value="be">беларуская</option><option value="bg">български</option><option value="ky">кыргызча</option><option value="kk">қазақ тілі</option><option value="mk">македонски</option><option value="mn">монгол</option><option value="ru">Русский</option><option value="sr">српски</option><option value="uk">Українська</option><option value="ka">ქართული</option><option value="hy">հայերեն</option><option value="iw">עברית</option><option value="ur">اردو</option><option value="ar">العربية</option><option value="fa">فارسی</option><option value="am">አማርኛ</option><option value="ne">नेपाली</option><option value="mr">मराठी</option><option value="hi">हिन्दी</option><option value="as">অসমীয়া</option><option value="bn">বাংলা</option><option value="pa">ਪੰਜਾਬੀ</option><option value="gu">ગુજરાતી</option><option value="or">ଓଡ଼ିଆ</option><option value="ta">தமிழ்</option><option value="te">తెలుగు</option><option value="kn">ಕನ್ನಡ</option><option value="ml">മലയാളം</option><option value="si">සිංහල</option><option value="th">ไทย</option><option value="lo">ລາວ</option><option value="my">မြန်မာ</option><option value="km">ខ្មែរ</option><option value="ko">한국어</option><option value="ja">日本語</option><option value="zh-CN">简体中文</option><option value="zh-TW">繁體中文</option><option value="zh-HK">繁體中文&nbsp;(香港)</option></select><input type="hidden" name="oldhl" value="en"><input type="hidden" name="gl" value="DE"><input type="hidden" name="m" value="0"><input type="hidden" name="app" value="0"><input type="hidden" name="pc" value="srp"><input type="hidden" name="continue" value="https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&amp;hl=en"><input type="hidden" name="src" value="1"><input type="hidden" name="cm" value="4"><input type="submit" value="Change language"></form><a href="https://policies.google.com/privacy?hl=en&utm_source=ucb">Privacy Policy</a> <a href="https://policies.google.com/terms?hl=en&utm_source=ucb">Terms of Service</a></div></div></body></html>
ERROR: Sorry, couldn't find any help with that topic
```
```
$ howdoi --version
WARNING:root:no boto3 module found
2.0.20
$ grep cookie $(locate howdoi.py)
                                  cookies={'CONSENT': 'YES+US.en+20170717-00-0'})
```

**Comment by gleitz** [comment]
How odd. The way your post renders in github for me is a page asking you to accept cookies

<img width="826" alt="Screenshot 2023-12-06 at 8 14 46 AM" src="https://github.com/gleitz/howdoi/assets/395254/9bc7c90a-cf5f-42e7-8296-a55f38598baa">

That query is working for me so perhaps you need to try some google searches in a browser and accept the cookie warning?

```
» howdoi bash case --explain -C
INFO: Version: 2.0.20
Cache cleared successfully
INFO: Fetching answers for query: bash case
INFO: Searching google with URL: https://www.google.com/search?q=site:stackoverflow.com%20bash%20case&hl=en
INFO: Links from stackoverflow.com found on google: 1
INFO: URL: https://stackoverflow.com/questions/16673570/bash-case-statement
INFO: Answers requested: 1, Starting at position: 1
INFO: Total answers returned: 1
#!/bin/sh

usage() {
    echo `basename $0`: ERROR: $* 1>&2
    echo usage: `basename $0` '[-a] [-b] [-c]
        [file ...]' 1>&2
    exit 1
}


while :
do
    case "$1" in
    -a|-A) echo you picked A;;
    -b|-B) echo you picked B;;
    -c|-C) echo you picked C;;
    -*) usage "bad argument $1";;
    *) break;;
    esac
    shift
done
```
